### PR TITLE
agent exec end to end tests

### PR DESCRIPTION
**Pull request by dhuckins**
- add tests
- use token file
- found another way to kill the cmd
- format
- add helper
- works
- add comment

### CODE REVIEW

File: command/agent_test.go [hunk 1]
```diff
@@ -3097,6 +3100,323 @@ vault {
 	wg.Wait()
 }
 
+func setupTestApp(t *testing.T) string {
```
Comment (averche) [inline command/agent_test.go:3103]:
```suggestion
func buildTestApp(t *testing.T) string {
```

File: command/agent_test.go [hunk 2]
```diff
@@ -3097,6 +3100,323 @@ vault {
...
+
+// TestAgent_Exec_Restarts tests that vault agent restarts the app
+// on changes to relevant vault secrets
+func TestAgent_Exec_Restarts(t *testing.T) {
+	testAppBin := setupTestApp(t)
+	defer os.Remove(testAppBin)
+
+	vaultClient, cleanup := testVaultServer(t)
+	defer cleanup()
+	t.Setenv(api.EnvVaultAddress, vaultClient.Address())
```
Comment (averche) [inline command/agent_test.go:3128]: perhaps add a TODO comment to remove this

File: command/agent_test.go [hunk 3]
```diff
@@ -3097,6 +3100,323 @@ vault {
...
+}
+
+auto_auth {
+	method {
+		type = "token_file"
+    	config {
+      		token_file_path = "%s"
+		}
+	}
+}
```
Comment (averche) [inline command/agent_test.go:3156]: Seems this might be a bit misformatted

File: command/agent_test.go [hunk 4]
```diff
@@ -3097,6 +3100,323 @@ vault {
...
+  contents = "{{ with secret \"kv-v2/data/my-db\" }}{{ .Data.data.user }}{{ end }}"
+}
+env_template "MY_DATABASE_PASSWORD" {
+  contents = "{{ with secret \"kv-v2/data/my-db\" }}{{ .Data.data.password }}{{ end }}"
+}
+
+exec {
+	command = ["%s", "-port", "%d"]
+}`, vaultClient.Address(), tokenFile.Name(), testAppBin, port)
+	logger := logging.NewVaultLogger(hclog.Trace)
```
Comment (averche) [inline command/agent_test.go:3167]: Nit: might want to initialize this at the top of the function

File: command/agent_test.go [hunk 5]
```diff
@@ -3097,6 +3100,323 @@ vault {
...
+
+	// wait until agent starts
+	select {
+	case <-agentCmd.startedCh:
+	case <-time.After(10 * time.Second):
+		t.Fatal("agent did not start within 10 seconds")
+	}
+
+	// agent started, give some time to populate env vars from vault
+	time.Sleep(10 * time.Second)
```
Comment (averche) [inline command/agent_test.go:3204]: 10s seems excessive. Maybe try retryablehttp to poll it like we did [in exec_test.go](https://github.com/hashicorp/vault/blob/main/command/agent/exec/exec_test.go#L300)?

File: command/agent_test.go [hunk 6]
```diff
@@ -3097,6 +3100,323 @@ vault {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	// wait for consul-template to fetch the new values
+	// see template_config.static_secret_render_interval
+	time.Sleep(4 * time.Second)
+
+	checkTestAppEnvVars(map[string]string{
+		"MY_DATABASE_USER":     "dbuser",
+		"MY_DATABASE_PASSWORD": "password1",
```
Comment (averche) [inline command/agent_test.go:3256]: Should this not be newuser/password2 instead?

File: command/agent_test.go [hunk 7]
```diff
@@ -3097,6 +3100,323 @@ vault {
...
+
+	// token file needs to have 600 permissions
+	if err := os.Chmod(tokenFile.Name(), 0o600); err != nil {
+		t.Fatalf("unable to change permissions of token file: %s", err)
+	}
+
+	const port = 34001
+	config := fmt.Sprintf(`
+template_config {
+  static_secret_render_interval = "2s"
```
Comment (averche) [inline command/agent_test.go:3141]:
```suggestion
  static_secret_render_interval = "1s"
```

File: command/agent_test.go (hunk 5 above)
Comment (dhuckins) [reply command/agent_test.go:3204]: its a little different... just waiting for consul-template to render. will change to something smaller tho

File: command/agent_test.go [hunk 8]
```diff
@@ -3097,6 +3100,326 @@ vault {
...
+}
+
+auto_auth {
+	method {
+		type = "token_file"
+    	config {
+			token_file_path = "%s"
+		}
+	}
+}
```
Comment (averche) [inline command/agent_test.go:3159]:
```suggestion
auto_auth {
  method {
    type = "token_file"
    config {
      token_file_path = "%s"
    }
  }
}
```

**Pull request closed without merging**